codesearch/index/index.go

```diff
@@ -228,28 +229,40 @@ func (w *Writer) lookupDocId(matchField types.Field) (uint64, error) {
 		return 0, status.InternalError("match field must be of keyword type")
 	}
 
-	var postingList posting.List
 	// First, check in the current batch
 	if pl, ok := w.fieldPostingLists[matchField.Name()][string(matchField.Contents())]; ok {
-		postingList = pl
-	} else {
-		// If not found, check in the index
-		key := postingListKey(w.namespace, string(matchField.Contents()), matchField.Name())
-		value, closer, err := w.db.Get(key)
-		if err != nil {
-			return 0, err
-		}
-		defer closer.Close()
+		return pl.ToArray()[0], nil
+	}
 
-		postingList, err = posting.Unmarshal(value)
-		if err != nil {
-			return 0, err
+	pl, closer, err := getPostingListReadOnly(w.db, w.namespace, string(matchField.Contents()), matchField.Name())
+	if err != nil {
+		return 0, err
+	}
+	defer closer.Close()
+
+	if pl.GetCardinality() > 1 {
+		return 0, status.FailedPreconditionErrorf("Match field matches > 1 (%d) docs: %v", pl.GetCardinality(), matchField)
+	} else if pl.GetCardinality() == 0 {
+		return 0, pebble.ErrNotFound // No matching documents
+	}
+	return pl.Iterator().Next(), nil
+}
+
+func getPostingListReadOnly(db pebble.Reader, namespace, key, field string) (posting.List, io.Closer, error) {
```

Comment:
another way to make this slightly safer would be to make getPostingListReadOnly a reader-specific method, and remove the changes on writer (idk how much they matter for perf?)

Reply:
My guess is the only place it really matters on the write-side is delete compaction, which is a very expensive operation. The others are looking up docIds, which should be posting lists of size 1. I'm still inclined to leave it as-is though - moving the reads into this helper got rid of a bunch of duplicated code.

Given that the Writer sometimes has to read, a cleaner factoring in general here might be a private `reader` or `dbwrapper` that has general, lower-level utility methods, and then have `Writer` and `Reader` both use that.